### ISSUE TITLE
Adapter injection applies a new adapter's inference mode to active adapters

### ISSUE THREAD

**Issue by DaoyuanLi2816** [huggingface/peft#3487]
### System Info

- PEFT `main` at `a5526d27a9d47d1e8264d5e1b1f96c0fdc79464e`
- PEFT 0.20.0
- Python 3.13.13
- PyTorch 2.13.0+cpu
- Transformers 5.14.1
- Windows 11

### Who can help?

@BenjaminBossan

### Reproduction

Four existing strict-xfail tests still reproduce the same adapter-specific trainability problem on current `main`:

```bash
pytest \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_set_correctly_switch_inference_mode \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_load_adapter_then_add_adapter \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_different_targets \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_same_targets \
  --runxfail -q
```

Result: 4 failed.

The failures are:

- Loading a frozen `default` adapter and then loading trainable `other` leaves the `default` parameters trainable instead of `other`.
- Adding another adapter to a model loaded with `is_trainable=False` re-enables gradients on the frozen adapter.
- Adding an AdaLoRA adapter with `inference_mode=True` freezes the still-active trainable adapter, for both same-target and different-target adapters.

These cases appear to share the housekeeping call in `BaseTuner.inject_adapter`:

```python
self.set_adapter(self.active_adapters, inference_mode=peft_config.inference_mode)
```

Here `peft_config` belongs to the adapter being injected, while `active_adapters` can refer to a different adapter. The new adapter's `inference_mode` is therefore applied across adapter boundaries.

If this scope is welcome, I can prepare one focused fix covering the load/add and AdaLoRA cases and turn the four existing xfails into regular regression tests. I would not add or change public API.

### Expected behavior

Injecting or loading an adapter should not change the trainability state of a different adapter. The active adapter should keep its current `requires_grad` state, while the newly loaded adapter should follow its own `is_trainable` or `inference_mode` setting.

**Comment by BenjaminBossan** [comment]
Thanks for bringing this up @DaoyuanLi2816.

We have a bit of a muddied behavior in PEFT where setting an adapter also changes the `requires_grad`, which can have unintended interactions. We will most likely decouple these concerns in the future. However, such a change would be backwards incompatible, so we reserve this for a 1.0 release.

In the meantime, if there is an easy fix for this problem which otherwise preserves the existing semantics, it's welcome. If this requires a complicated fix, I'd rather stay away from it and wait for the 1.0 release, which should address it.

**Comment by DaoyuanLi2816** [comment]
Thanks. I have a focused implementation and regression coverage ready locally: the four reproductions pass, and the broader requires-grad slice passes as well (210 passed, 4 skipped).

One of the four cases—loading a trainable adapter while the active adapter is frozen—touches broader `load_adapter` trainability semantics, so I can keep that separate if you prefer the narrower semantics-preserving fix.

I have not opened a PR because #3488 is still open and the contribution policy asks contributors not to open overlapping PRs. I still intend to submit the fix discussed here. Should I wait for #3488 to be closed, or would you prefer that I open mine now for comparison?

**Comment by BenjaminBossan** [comment]
Please go ahead with your PR, you have precedence over #3488 since you proposed to open the PR earlier.

> One of the four cases—loading a trainable adapter while the active adapter is frozen—touches broader `load_adapter` trainability semantics, so I can keep that separate if you prefer the narrower semantics-preserving fix.

Let's create the PR without that change. In the PR description, please sketch the change required in `load_adapter` so that we can check if it's a worthwhile tradeoff.

**Comment by BenjaminBossan** [comment]
_not stale_
